### ISSUE TITLE
encoder.flush() hangs in lock_mapping() mutex

### ISSUE THREAD

**Issue by liquidnight2** [gfx-rs/gfx#1164]
This program runs until the last few lines, and stops then with 0% cpu load.

File:  Cargo.toml:
```toml
 [dependencies]
 winit = "*"
 gfx_core = "0.6.0"
 gfx = "0.14.0"
 gfx_window_glutin = "0.14.0"
 glutin = "*"
 gfx_device_gl = "*"
```

File:  shader/pos_scale_trans_150.glslv
```glsl
#version 150 core
in vec3 a_Color;
in vec2 a_Position;
in vec2 a_Translate;
uniform float u_Scale;
out vec4 v_Color;
void main() {
   v_Color = vec4(a_Color, 1.0);
   gl_Position = vec4(a_Position*u_Scale + a_Translate  , 0.0, 1.0);
}
```

File:   shader/triangle_150.glslf
```glsl
#version 150 core
in vec4 v_Color;
out vec4 Target0;
void main() {
    Target0 = v_Color;
}
```
Program:
```rust
extern crate winit;
extern crate glutin;
extern crate gfx_core;
#[macro_use]
extern crate gfx;
extern crate gfx_window_glutin;
extern crate gfx_device_gl;

use gfx::traits::FactoryExt;
use gfx::traits::Factory;
use std::time::Duration;
use std::thread;
use gfx::Device;
use gfx::pso::PipelineState;

pub type ColorFormat = gfx::format::Rgba8;
pub type DepthFormat = gfx::format::DepthStencil;

gfx_defines!{
    vertex Vertex2D {
        pos: [f32; 2] = "a_Position",
    }

    vertex Instance {
        translate: [f32; 2] = "a_Translate",
        color: [f32; 3] = "a_Color",
    }

    constant Locals {
        scale: f32 = "u_Scale",
    }

    pipeline pipe_ma {
        vertexbuffer: gfx::VertexBuffer<Vertex2D> = (),
        instance: gfx::InstanceBuffer<Instance> = (),
        scale: gfx::Global<f32> = "u_Scale",
        locals: gfx::ConstantBuffer<Locals> = "Locals",
        out: gfx::RenderTarget<ColorFormat> = "Target0",
    }
}

const COLOR_RED: [f32; 3] = [1.0, 0.1, 0.1];
const COLOR_BLUE: [f32; 3] = [0.1, 0.1, 1.0];
const COLOR4_CLEAR: [f32; 4] = [0.1, 0.1, 0.1, 1.0];

pub fn make_square(px: f32, py: f32, size: f32) -> ([Vertex2D; 4], [u16; 6]) {
    ([Vertex2D { pos: [px - size, py - size] },
      Vertex2D { pos: [px + size, py - size] },
      Vertex2D { pos: [px + size, py + size] },
      Vertex2D { pos: [px - size, py + size] }],
     [0, 1, 2, 2, 3, 0])
}

fn create_upload_buffer<R, F>(fac: &mut F, instance_count: u32) -> gfx::handle::Buffer<R, Instance>
    where F: gfx::Factory<R> + gfx::traits::FactoryExt<R>,
          R: gfx::Resources
{
    let ubuf: gfx::handle::Buffer<R, Instance> = fac.create_upload_buffer(instance_count as usize)
        .unwrap();
    ubuf
}


pub fn main() {
    let instance_count: u32 = 2;

    let builder = glutin::WindowBuilder::new()
        .with_title("Double Square 3".to_string())
        .with_dimensions(1024, 768);
    let (window, mut device, mut factory1, window_main_color, mut _main_depth) 
       : (glutin::Window, gfx_device_gl::Device  , gfx_device_gl::Factory,
        gfx_core::handle::RenderTargetView<gfx_device_gl::Resources,
           (gfx_core::format::R8_G8_B8_A8, gfx_core::format::Unorm)> ,
        gfx_core::handle::DepthStencilView<gfx_device_gl::Resources,
           (gfx_core::format::D24_S8, gfx_core::format::Unorm)>  )
    = gfx_window_glutin::init::<ColorFormat, DepthFormat>(builder);
    let vertexshader: &[u8] = include_bytes!("shader/pos_scale_trans_150.glslv");
    let pixelshader = include_bytes!("shader/triangle_150.glslf");
    let pipeline_state_object: PipelineState<_, _> = factory1.create_pipeline_simple(//
                                vertexshader,
                                pixelshader,
                                pipe_ma::new())
        .unwrap();
    let (square_vertexes, square_indices) = make_square(-0.0, 0.1, 0.2);
    let uploadbuffer: gfx_core::handle::Buffer<gfx_device_gl::Resources, Instance> =
        create_upload_buffer(&mut factory1, instance_count);
    let mut writer = factory1.write_mapping(&uploadbuffer).unwrap();
    writer[0] = Instance {
        translate: [0.1, 0.0],
        color: COLOR_BLUE,
    };
    writer[1] = Instance {
        translate: [-0.1, 0.1],
        color: COLOR_RED,
    };
    let instancebuffer = factory1.create_buffer(instance_count as usize,
                       gfx::buffer::Role::Vertex,
                       gfx::memory::Usage::Data,
                       gfx::TRANSFER_DST)
        .unwrap();
    let (vertex_buffer, mut slice) =
        factory1.create_vertex_buffer_with_slice(&square_vertexes as &[Vertex2D],
                                                 &square_indices as &[u16]);
    slice.instances = Some((instance_count, 0));
    let locals1 = Locals { scale: 1.0 };
    let localsbuffer =
        factory1.create_buffer_immutable(&[locals1],
            gfx::buffer::Role::Constant, gfx::Bind::empty())
            .unwrap();
    let mut data = pipe_ma::Data {
        vertexbuffer: vertex_buffer,
        out: window_main_color,
        instance: instancebuffer,
        locals: localsbuffer,
        scale: 1.0,
    };

    let mut encoder: gfx::Encoder<_, _> = factory1.create_command_buffer().into();
    encoder.copy_buffer(&uploadbuffer, &data.instance, 0, 0, instance_count as usize)
        .unwrap();
    'main: loop {
        for event in window.poll_events() {
            match event {
                glutin::Event::KeyboardInput(_, _, Some(glutin::VirtualKeyCode::Escape)) |
                glutin::Event::Closed => break 'main,
                glutin::Event::Resized(_width, _height) => {
                    gfx_window_glutin::update_views(&window, &mut data.out, &mut _main_depth);
                }
                _ => {}
            }
        }
        println!("encoder clear ...");
        encoder.clear(&data.out, COLOR4_CLEAR);
        println!("encoder draw ...");
        encoder.draw(&slice, &pipeline_state_object, &data);
        println!("encoder flush ...");
        encoder.flush(&mut device);
        println!("encoder swap ...");
        window.swap_buffers().unwrap();
        println!("dev cleanup  ...");
        device.cleanup();
        println!("wait ...");
        thread::sleep(Duration::from_millis(200));
    }
    println!("END!");
}
```

Call Stack from gdb:
```
Thread 1 "suspectedlock" received signal SIGINT, Interrupt.
0x00007ffff79c830c in __lll_lock_wait () from /lib64/libpthread.so.0
(gdb) bt
#0  0x00007ffff79c830c in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007ffff79c1e15 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00005555555ed83a in std::sys::imp::mutex::Mutex::lock (self=0x7ffff6d80d10)
    at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libstd/sys/unix/mutex.rs:67
#3  0x00005555555ebb9d in std::sys_common::mutex::Mutex::lock (self=0x7ffff6d80d10)
    at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libstd/sys_common/mutex.rs:40
#4  0x00005555555f39bb in std::sync::mutex::Mutex<gfx_device_gl::factory::MappingGate>::lock<gfx_device_gl::factory::MappingGate> (self=0x7ffff6c5b1f0)
    at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libstd/sync/mutex.rs:189
#5  0x00005555556441ec in gfx_core::buffer::{{impl}}::lock_mapping::{{closure}}<gfx_device_gl::Resources> (m=0x7ffff6c5b1f0)
    at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_core-0.6.0/src/buffer.rs:57
#6  0x00005555555dc270 in core::option::Option<&std::sync::mutex::Mutex<gfx_device_gl::factory::MappingGate>>::map<&std::sync::mutex::Mutex<gfx_device_gl::factory::MappingGate>,std::sync::mutex::MutexGuard<gfx_device_gl::factory::MappingGate>,closure> (self=..., f=...)
    at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libcore/option.rs:383
#7  0x00005555555eb950 in gfx_core::buffer::Raw<gfx_device_gl::Resources>::lock_mapping<gfx_device_gl::Resources> (self=0x7ffff6c5b1d0)
    at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_core-0.6.0/src/buffer.rs:57
#8  0x0000555555640735 in gfx_device_gl::Device::ensure_mappings_flushed (self=0x7fffffffd1d0, buffers=...)
    at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_device_gl-0.13.0/src/lib.rs:779
#9  0x00005555556405a9 in gfx_device_gl::Device::before_submit (self=0x7fffffffd1d0, gpu_access=0x7fffffffc458)
    at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_device_gl-0.13.0/src/lib.rs:767
#10 0x00005555556413b9 in gfx_device_gl::{{impl}}::submit (self=0x7fffffffd1d0, cb=0x7fffffffbf78, access=0x7fffffffc458)
    at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx_device_gl-0.13.0/src/lib.rs:885
#11 0x0000555555576f70 in gfx::encoder::Encoder<gfx_device_gl::Resources, gfx_device_gl::command::CommandBuffer>::flush<gfx_device_gl::Resources,gfx_device_gl::command::CommandBuffer,gfx_device_gl::Device> (self=0x7fffffffbf78, device=0x7fffffffd1d0) at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/gfx-0.14.0/src/encoder.rs:174
#12 0x00005555555915da in suspectedlock::main () at /home/user/workspacer/newProj3/src/main.rs:183
#13 0x00005555557983bb in panic_unwind::__rust_maybe_catch_panic () at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libpanic_unwind/lib.rs:97
#14 0x000055555578fcc6 in std::panicking::try<(),fn()> () at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libstd/panicking.rs:332
#15 std::panic::catch_unwind<fn(),()> () at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libstd/panic.rs:351
#16 std::rt::lang_start () at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/obj/../src/libstd/rt.rs:57
#17 0x000055555559a573 in main ()
```

Apitrace dump:
```
2 glXChooseFBConfig(dpy = 0x7fc95926d000, screen = 0, attribList = {GLX_X_RENDERABLE, True, GLX_X_VISUAL_TYPE, GLX_TRUE_COLOR, GLX_DRAWABLE_TYPE, GLX_WINDOW_BIT, GLX_RENDER_TYPE, GLX_RGBA_BIT, GLX_RED_SIZE, 8, GLX_GREEN_SIZE, 8, GLX_BLUE_SIZE, 8, GLX_ALPHA_SIZE, 8, GLX_DEPTH_SIZE, 24, GLX_STENCIL_SIZE, 8, GLX_DOUBLEBUFFER, True, GLX_STEREO, False, GLX_CONFIG_CAVEAT, -1, 0}, nitems = &20) = {0x7fc954fbc400, 0x7fc954fbc5c0, 0x7fc9551c0d00, 0x7fc9551c0ec0, 0x7fc9551bc1c0, 0x7fc9551bc460, 0x7fc9551c2ac0, 0x7fc9551c2d60, 0x7fc9551bc2a0, 0x7fc9551bc540, 0x7fc9551c2ba0, 0x7fc9551c2e40, 0x7fc9551bc380, 0x7fc9551bc620, 0x7fc9551c2c80, 0x7fc9551c2f20, 0x7fc954fbc4e0, 0x7fc954fbc6a0, 0x7fc9551c0de0, 0x7fc9551c0fa0}
15 glXGetVisualFromFBConfig(dpy = 0x7fc95926d000, config = 0x7fc954fbc400) = &{visual = 0x7fc9592a61c0, visualid = 33, screen = 0, depth = 24, c_class = 4, red_mask = 16711680, green_mask = 65280, blue_mask = 255, colormap_size = 256, bits_per_rgb = 8}
58 glXCreateContextAttribsARB(dpy = 0x7fc95926d000, config = 0x7fc954fbc400, share_context = NULL, direct = True, attrib_list = {GLX_CONTEXT_MAJOR_VERSION_ARB, 3, GLX_CONTEXT_MINOR_VERSION_ARB, 2, GLX_CONTEXT_FLAGS_ARB, GLX_CONTEXT_DEBUG_BIT_ARB, 0}) = 0x7fc95924ca40
59 glXMakeCurrent(dpy = 0x7fc95926d000, drawable = 48234498, ctx = 0x7fc95924ca40) = True
60 glViewport(x = 0, y = 0, width = 634, height = 434)
61 glScissor(x = 0, y = 0, width = 634, height = 434)
765 glGetIntegerv(pname = GL_NUM_EXTENSIONS, params = &183)
949 glGetIntegerv(pname = GL_MAX_ELEMENTS_VERTICES, params = &3000)
950 glGetIntegerv(pname = GL_MAX_ELEMENTS_INDICES, params = &3000)
951 glGetIntegerv(pname = GL_MAX_TEXTURE_SIZE, params = &16384)
952 glGetIntegerv(pname = GL_MAX_PATCH_VERTICES, params = &32)
953 glEnable(cap = GL_FRAMEBUFFER_SRGB)
954 glPixelStorei(pname = GL_UNPACK_ALIGNMENT, param = 1)
955 glEnable(cap = GL_PROGRAM_POINT_SIZE)
956 glGenVertexArrays(n = 1, arrays = &1)
957 glBindVertexArray(array = 1)
959 glCreateShader(type = GL_VERTEX_SHADER) = 1
960 glShaderSource(shader = 1, count = 1, string = &"#version 150 core
in vec3 a_Color;
in vec2 a_Position;
in vec2 a_Translate;

uniform float u_Scale;

out vec4 v_Color;

void main() {
   v_Color = vec4(a_Color, 1.0);
   gl_Position = vec4(a_Position*u_Scale + a_Translate  , 0.0, 1.0);
}
", length = &238)
961 glCompileShader(shader = 1)
962 glGetShaderiv(shader = 1, pname = GL_COMPILE_STATUS, params = &1)
963 glGetShaderiv(shader = 1, pname = GL_INFO_LOG_LENGTH, params = &0)
964 glCreateShader(type = GL_FRAGMENT_SHADER) = 2
965 glShaderSource(shader = 2, count = 1, string = &"#version 150 core

in vec4 v_Color;
out vec4 Target0;

void main() {
    Target0 = v_Color;
}
", length = &94)
966 glCompileShader(shader = 2)
967 glGetShaderiv(shader = 2, pname = GL_COMPILE_STATUS, params = &1)
968 glGetShaderiv(shader = 2, pname = GL_INFO_LOG_LENGTH, params = &0)
969 glCreateProgram() = 3
970 glAttachShader(program = 3, shader = 1)
971 glAttachShader(program = 3, shader = 2)
972 glBindAttribLocation(program = 3, index = 0, name = "a_Color")
973 glBindAttribLocation(program = 3, index = 1, name = "a_Position")
974 glBindAttribLocation(program = 3, index = 2, name = "a_Translate")
975 glLinkProgram(program = 3)
976 glGetProgramiv(program = 3, pname = GL_LINK_STATUS, params = &1)
977 glGetProgramiv(program = 3, pname = GL_INFO_LOG_LENGTH, params = &0)
978 glGetProgramiv(program = 3, pname = GL_ACTIVE_UNIFORMS, params = &1)
979 glGetActiveUniformsiv(program = 3, uniformCount = 1, uniformIndices = &0, pname = GL_UNIFORM_BLOCK_INDEX, params = 0x7fc959215258)
980 glGetActiveUniformsiv(program = 3, uniformCount = 1, uniformIndices = &0, pname = GL_UNIFORM_OFFSET, params = 0x7fc959215238)
981 glGetProgramiv(program = 3, pname = GL_ACTIVE_UNIFORM_MAX_LENGTH, params = &8)
982 glUseProgram(program = 3)
983 glGetActiveUniform(program = 3, index = 0, bufSize = 8, length = &7, size = &1, type = &GL_FLOAT, name = "u_Scale")
984 glGetUniformLocation(program = 3, name = "u_Scale") = 0
985 glGetProgramiv(program = 3, pname = GL_ACTIVE_ATTRIBUTES, params = &3)
986 glGetProgramiv(program = 3, pname = GL_ACTIVE_ATTRIBUTE_MAX_LENGTH, params = &12)
987 glGetActiveAttrib(program = 3, index = 0, bufSize = 12, length = &7, size = &1, type = &GL_FLOAT_VEC3, name = "a_Color")
988 glGetAttribLocation(program = 3, name = "a_Color") = 0
989 glGetActiveAttrib(program = 3, index = 1, bufSize = 12, length = &10, size = &1, type = &GL_FLOAT_VEC2, name = "a_Position")
990 glGetAttribLocation(program = 3, name = "a_Position") = 1
991 glGetActiveAttrib(program = 3, index = 2, bufSize = 12, length = &11, size = &1, type = &GL_FLOAT_VEC2, name = "a_Translate")
992 glGetAttribLocation(program = 3, name = "a_Translate") = 2
993 glGetProgramiv(program = 3, pname = GL_ACTIVE_UNIFORM_BLOCKS, params = &0)
994 glGetProgramiv(program = 3, pname = GL_ACTIVE_UNIFORM_MAX_LENGTH, params = &8)
995 glGetProgramInterfaceiv(program = 3, programInterface = GL_PROGRAM_OUTPUT, pname = GL_ACTIVE_RESOURCES, params = &1)
996 glGetProgramResourceiv(program = 3, programInterface = GL_PROGRAM_OUTPUT, index = 0, propCount = 1, props = &GL_NAME_LENGTH, bufSize = 1, length = NULL, params = &8)
997 glGetProgramResourceName(program = 3, programInterface = GL_PROGRAM_OUTPUT, index = 0, bufSize = 8, length = NULL, name = "Target0")
998 glGetProgramResourceiv(program = 3, programInterface = GL_PROGRAM_OUTPUT, index = 0, propCount = 1, props = &GL_LOCATION, bufSize = 1, length = NULL, params = &0)
999 glGetProgramResourceiv(program = 3, programInterface = GL_PROGRAM_OUTPUT, index = 0, propCount = 1, props = &GL_TYPE, bufSize = 1, length = NULL, params = &35666)
1001 glGenBuffers(n = 1, buffer = &1)
1002 glBindBuffer(target = GL_ARRAY_BUFFER, buffer = 1)
1003 glBufferStorage(target = GL_ARRAY_BUFFER, size = 40, data = NULL, flags = GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT)
1005 glBindBuffer(target = GL_ARRAY_BUFFER, buffer = 1)
1006 glMapBufferRange(target = GL_ARRAY_BUFFER, offset = 0, length = 40, access = GL_MAP_WRITE_BIT | GL_MAP_FLUSH_EXPLICIT_BIT | GL_MAP_PERSISTENT_BIT) = 0x7fc95aaf9000
1008 glGenBuffers(n = 1, buffer = &2)
1009 glBindBuffer(target = GL_ARRAY_BUFFER, buffer = 2)
1010 glBufferStorage(target = GL_ARRAY_BUFFER, size = 40, data = NULL, flags = 0x0)
1012 glGenBuffers(n = 1, buffer = &3)
1013 glBindBuffer(target = GL_ARRAY_BUFFER, buffer = 3)
1014 glBufferStorage(target = GL_ARRAY_BUFFER, size = 32, data = blob(32), flags = 0x0)
1016 glGenBuffers(n = 1, buffer = &4)
1017 glBindBuffer(target = GL_ELEMENT_ARRAY_BUFFER, buffer = 4)
1018 glBufferStorage(target = GL_ELEMENT_ARRAY_BUFFER, size = 12, data = blob(12), flags = 0x0)
1020 glGenBuffers(n = 1, buffer = &5)
1021 glBindBuffer(target = GL_UNIFORM_BUFFER, buffer = 5)
1022 glBufferStorage(target = GL_UNIFORM_BUFFER, size = 4, data = blob(4), flags = 0x0)
1024 glGenFramebuffers(n = 1, framebuffers = &1)
```

**Comment by msiglreith** [comment]
Probably caused by permanently mapping the buffer here
`let mut writer = factory1.write_mapping(&uploadbuffer).unwrap();`

Putting a scope around it
```Rust
{
      let mut writer = factory1.write_mapping(&uploadbuffer).unwrap();
        writer[0] = Instance {
            translate: [0.1, 0.0],
            color: COLOR_BLUE,
        };
        writer[1] = Instance {
            translate: [-0.1, 0.1],
            color: COLOR_RED,
        };  
    }
```
might fix it.

**Comment by liquidnight2** [comment]
msiglreith: Yes, that did help.  

Where do I find that requirement in the documentation ?

**Comment by kvark** [comment]
@liquidnight2 it's still our issue, so the documentation shouldn't be changed, I think.
We should make sure that the case is handled. Any crash and or hang is not your fault.

**Comment by Bastacyclop** [comment]
At first I used `try_lock` and crashed on failure, reporting an user error. This is great for OpenGL since all is single-threaded, but when mapping access (including the implicit submission access) may be done on different threads, this is very strict. That's why I used `lock` instead, so that multi-thread accesses could result in a stall instead of a crash. The possiblities are:
- be strict and use `try_lock`, any access overlap is an error, the user will need its own runtime synchronization if he cannot statically guarantee no overlap.
- allow some access overlap with `lock` at the cost of implicit stalling and potential deadlocks.
- something else I did not think about

Anyway the documentation should be more clear about the adopted behavior.
After seeing how easy it is to create a deadlock and considering the safety goal of GFX, I tend to be in favor of the strict way.
